tamr_client/datasets/dataframe.py

```diff
@@ -0,0 +1,59 @@
+"""
+Convenient functionality for interacting with pandas DataFrames.
+"""
+
+import json
+from typing import Optional
+
+import pandas as pd
+
+import tamr_client as tc
+from tamr_client.types import JsonDict
+
+
+def upsert(
+    session: tc.Session,
+    dataset: tc.Dataset,
+    df: pd.DataFrame,
+    *,
+    primary_key_name: Optional[str] = None,
+) -> JsonDict:
+    """Upserts a record for each row of `df` with attributes for each column in `df`.
+    Args:
+        dataset: Dataset to receive record updates
+        df: The DataFrame containing records to be upserted
+        primary_key_name: The primary key of the dataset.  Must be a column of `df`. By default the key_attribute_name of dataset
+
+    Returns:
+        JSON response body from the server
+
+    Raises:
+        requests.HTTPError: If an HTTP error is encountered
+        PrimaryKeyNotFound: If `primary_key_name` is not a column in `df` or the index of `df`
+        ValueError: If `primary_key_name` matches both a column in `df` and the index of `df`
+    """
+    if not primary_key_name:
```

Comment:
Technically, `pd.DataFrame` columns can have empty string `''` as a column name, so its better to do an explicit check for `None` (`not primary_key_name` will be `True` for the empty string since empty strings are false-y in Python).

```suggestion
    if primary_key_name is None:
```